library/core/src/cmp.rs

```diff
@@ -1161,7 +1175,11 @@ pub trait PartialOrd<Rhs: ?Sized = Self>: PartialEq<Rhs> {
     #[must_use]
     #[stable(feature = "rust1", since = "1.0.0")]
     fn gt(&self, other: &Rhs) -> bool {
-        matches!(self.partial_cmp(other), Some(Greater))
+        if let Some(ordering) = self.partial_cmp(other) {
+            ordering.is_gt()
```

Comment:
This is now conceptually two checks, rather than just one, so it's possible it's not always better.  [None is currently 2](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=baf920f8691f7409c76f924d810ec436) here, so the old code was hypothetically just `c == 1`, and now it's `c != 2 && c > 0`.  (Of course `lt` ends up being `c != 2 && c < 0`, which obviously folds to `c < 0`, so that one's probably not impacted.)

My hope is that this is still better in practice:
1. I would bet that most `partial_cmp`s are actually `cmp`s, and thus the optimizer will easily notice that the result is never `None` -- like happens in the codegen test.
2. For things that *can* actually return `None`, hopefully jump-threading will usually notice that the `None` becomes `false` and will again bypass actually running this check at runtime.

I'll see if I can prove that out in a codegen test...

Reply:
Well, I didn't manage to make a great codegen test for this, but I did in passing find two other things:
- We should start putting more `noundef` on parameters, <https://rust-lang.zulipchat.com/#narrow/stream/187780-t-compiler.2Fwg-llvm/topic/We.20will.20want.20a.20lot.20of.20noundefs/near/317472833>
- LLVM doesn't optimize everything as well as it should: https://github.com/llvm/llvm-project/issues/59666